Can't double arbitrary callables on a instance of a class

Hello,

It feels like the following check in `doubles.verification.verify_method`:

```python
    if attr.kind == 'data' and not isbuiltin(attr.object):
        raise VerifyingDoubleError(method_name, target.doubled_obj).not_callable()
```

Could be replaced by:

```python
    if attr.kind == 'data' and not isbuiltin(attr.object) and not _is_callable(attr.object):
        raise VerifyingDoubleError(method_name, target.doubled_obj).not_callable()
```

(`_is_callable` being defined in `doubles.target`).

To allow this use case to work (as expected?):

```python
from doubles import expect


class Callable(object):

    def __call__(self):
        return u"Hello"


class Foo(object):

    operation = Callable()

t = Foo()


def test_operation():
    expect(t).operation.with_no_args().and_return(u"Bye bye").once()
    assert t.operation() == u"Bye bye"

```

Which currently fails with:

```
================================================================ test session starts ================================================================
platform linux2 -- Python 2.7.11, pytest-2.8.7, py-1.4.31, pluggy-0.3.1
rootdir: /tmp/louis/tmp, inifile: 
plugins: doubles-1.1.3
collected 1 items 

test.py F

===================================================================== FAILURES ======================================================================
__________________________________________________________________ test_operation ___________________________________________________________________

    def test_operation():
>       expect(t).operation.with_no_args().and_return(u"Bye bye").once()

test.py:18: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/home/louis/envs/scratch/lib/python2.7/site-packages/doubles/targets/expectation_target.py:67: in __getattribute__
    return self._proxy.add_expectation(attr_name, caller)
/home/louis/envs/scratch/lib/python2.7/site-packages/doubles/proxy.py:37: in add_expectation
    return self.method_double_for(method_name).add_expectation(caller)
/home/louis/envs/scratch/lib/python2.7/site-packages/doubles/proxy.py:67: in method_double_for
    self._method_doubles[method_name] = MethodDouble(method_name, self._target)
/home/louis/envs/scratch/lib/python2.7/site-packages/doubles/method_double.py:21: in __init__
    self._verify_method()
/home/louis/envs/scratch/lib/python2.7/site-packages/doubles/method_double.py:152: in _verify_method
    verify_method(self._target, self._method_name, class_level=class_level)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

target = <doubles.target.Target object at 0x7ffbba4ffd90>, method_name = 'operation', class_level = False

    def verify_method(target, method_name, class_level=False):
        """Verifies that the provided method exists on the target object.
    
        :param Target target: A ``Target`` object containing the object with the method to double.
        :param str method_name: The name of the method to double.
        :raise: ``VerifyingDoubleError`` if the attribute doesn't exist, if it's not a callable object,
            and in the case where the target is a class, that the attribute isn't an instance method.
        """
    
        attr = target.get_attr(method_name)
    
        if not attr:
            raise VerifyingDoubleError(method_name, target.doubled_obj).no_matching_method()
    
        if attr.kind == 'data' and not isbuiltin(attr.object):
>           raise VerifyingDoubleError(method_name, target.doubled_obj).not_callable()
E           VerifyingDoubleError: Cannot double method 'operation' because it is not a callable attribute on <test.Foo object at 0x7ffbba4ff8d0>.

/home/louis/envs/scratch/lib/python2.7/site-packages/doubles/verification.py:85: VerifyingDoubleError
==================================================== 1 failed, 1 pytest-warnings in 0.01 seconds ====================================================
```

I actually just tried that with the same examples and it would require more changes:

```
================================================================ test session starts ================================================================
platform linux2 -- Python 2.7.11, pytest-2.8.7, py-1.4.31, pluggy-0.3.1
rootdir: /tmp/louis/tmp, inifile: 
plugins: doubles-1.1.3
collected 1 items 

test.py F

===================================================================== FAILURES ======================================================================
__________________________________________________________________ test_operation ___________________________________________________________________

    def test_operation():
>       expect(t).operation.with_no_args().and_return(u"Bye bye").once()

test.py:18: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/home/louis/envs/scratch/lib/python2.7/site-packages/doubles/allowance.py:184: in with_no_args
    self.verify_arguments()
/home/louis/envs/scratch/lib/python2.7/site-packages/doubles/allowance.py:258: in verify_arguments
    verify_arguments(self._target, self._method_name, args, kwargs)
/home/louis/envs/scratch/lib/python2.7/site-packages/doubles/verification.py:120: in verify_arguments
    _verify_arguments(method, method_name, args, kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

method = <test.Callable object at 0x7f3c2d6f0910>, method_name = 'operation', args = ['self_or_cls'], kwargs = {}

    def _verify_arguments(method, method_name, args, kwargs):
        try:
            getcallargs(method, *args, **kwargs)
        except TypeError as e:
            if not _is_python_function(method):
>               raise VerifyingBuiltinDoubleArgumentError(str(e))
E               VerifyingBuiltinDoubleArgumentError: <test.Callable object at 0x7f3c2d6f0910> is not a Python function

/home/louis/envs/scratch/lib/python2.7/site-packages/doubles/verification.py:128: VerifyingBuiltinDoubleArgumentError
==================================================== 1 failed, 1 pytest-warnings in 0.01 seconds ====================================================
```

--- a/doubles/target.py
+++ b/doubles/target.py
@@ -1,15 +1,10 @@
-from inspect import classify_class_attrs, isclass, ismodule, isfunction, getmembers
+from inspect import classify_class_attrs, isclass, ismodule, getmembers
 from collections import namedtuple
 
 from doubles.object_double import ObjectDouble
+from doubles.verification import is_callable
 
 Attribute = namedtuple('Attribute', ['object', 'kind', 'defining_class'])
-
-
-def _is_callable(obj):
-    if isfunction(obj):
-        return True
-    return hasattr(obj, '__call__')
 
 
 def _proxy_class_method_to_instance(original, name):
@@ -90,7 +85,7 @@ class Target(object):
         attrs = {}
 
         if ismodule(self.doubled_obj):
-            for name, func in getmembers(self.doubled_obj, _is_callable):
+            for name, func in getmembers(self.doubled_obj, is_callable):
                 attrs[name] = Attribute(func, 'toplevel', self.doubled_obj)
         else:
             for attr in classify_class_attrs(self.doubled_obj_type):
@@ -151,7 +146,7 @@ class Target(object):
             return None
 
         func = getattr(self.doubled_obj, attr_name)
-        if not _is_callable(func):
+        if not is_callable(func):
             return None
 
         attr = Attribute(

--- a/doubles/testing.py
+++ b/doubles/testing.py
@@ -6,11 +6,20 @@ class EmptyClass(OldStyleEmptyClass, object):
     pass
 
 
+class ArbitraryCallable(object):
+    def __init__(self, value):
+        self.value = value
+
+    def __call__(self):
+        return self.value
+
+
 class OldStyleUser():
     """An importable dummy class used for testing purposes."""
 
     class_attribute = 'foo'
     callable_class_attribute = classmethod(lambda cls: 'dummy result')
+    arbitrary_callable = ArbitraryCallable('ArbitraryCallable Value')
 
     def __init__(self, name, age):
         self.name = name

--- a/doubles/verification.py
+++ b/doubles/verification.py
@@ -25,6 +25,12 @@ def _is_python_function(func):
     return isfunction(func)
 
 
+def is_callable(obj):
+    if isfunction(obj):
+        return True
+    return hasattr(obj, '__call__')
+
+
 def _is_python_33():
     v = sys.version_info
     return v[0] == 3 and v[1] == 3
@@ -81,7 +87,7 @@ def verify_method(target, method_name, class_level=False):
     if not attr:
         raise VerifyingDoubleError(method_name, target.doubled_obj).no_matching_method()
 
-    if attr.kind == 'data' and not isbuiltin(attr.object):
+    if attr.kind == 'data' and not isbuiltin(attr.object) and not is_callable(attr.object):
         raise VerifyingDoubleError(method_name, target.doubled_obj).not_callable()
 
     if class_level and attr.kind == 'method' and method_name != '__new__':
@@ -104,7 +110,7 @@ def verify_arguments(target, method_name, args, kwargs):
     attr = target.get_attr(method_name)
     method = attr.object
 
-    if attr.kind in ('attribute', 'toplevel', 'class method', 'static method'):
+    if attr.kind in ('data', 'attribute', 'toplevel', 'class method', 'static method'):
         try:
             method = method.__get__(None, attr.defining_class)
         except AttributeError:

--- a/test/partial_double_test.py
+++ b/test/partial_double_test.py
@@ -13,6 +13,19 @@ import doubles.testing
 
 @mark.parametrize('test_class', [User, OldStyleUser])
 class TestInstanceMethods(object):
+    def test_arbitrary_callable_on_instance(self, test_class):
+        instance = test_class('Bob', 10)
+        allow(instance).arbitrary_callable.and_return('Bob Barker')
+        assert instance.arbitrary_callable() == 'Bob Barker'
+        teardown()
+        assert instance.arbitrary_callable() == 'ArbitraryCallable Value'
+
+    def test_arbitrary_callable_on_class(self, test_class):
+        allow(test_class).arbitrary_callable.and_return('Bob Barker')
+        assert test_class.arbitrary_callable() == 'Bob Barker'
+        teardown()
+        assert test_class.arbitrary_callable() == 'ArbitraryCallable Value'
+
     def test_callable_class_attribute(self, test_class):
         allow(test_class).callable_class_attribute.and_return('Bob Barker')
         assert test_class.callable_class_attribute() == 'Bob Barker'

